internal/targets/hash_bins.go

```diff
@@ -0,0 +1,82 @@
+package targets
+
+import (
+	"strconv"
+	"strings"
+)
+
+// hexEncode formats x as a hex string, left padded with zeros to padWidth.
+func hexEncode(x uint64, padWidth int) string {
+	// Benchmarked to be more than 10x faster than padding with Sprintf.
+	s := strconv.FormatUint(x, 16)
+	if len(s) >= padWidth {
+		return s
+	}
+	return strings.Repeat("0", padWidth-len(s)) + s
+}
+
+// HashBin represents a hex prefix range. First should be less than Last.
+type HashBin struct {
+	First uint64
+	Last  uint64
+}
+
+// Name returns the of the role that signs for the HashBin.
+func (b HashBin) Name(prefix string, padWidth int) string {
```

Comment:
Depending on who can call this, you might want to validate that padWidth is positive

Reply:
I believe the code is safe against negative `padWidth`. If you pass a negative value, then the `len(s) >= padWidth` check in `hexEncode` does not apply any extra padding (same as if `padWidth == 0` or `padWidth < len(s)`).

```go
func hexEncode(x uint64, padWidth int) string {
	// ...
	if len(s) >= padWidth {
		return s
	}
	return strings.Repeat("0", padWidth-len(s)) + s
}
```